Update writing policies guide (#1195)
osohq/oso: docs/content/any/guides/policies/index.md

---

> title: Write Custom Policies

Reviewer: Consider naming Write Rules?

Author: ```suggestion
title: Write Rules
```

description: Learn about writing Oso policies - the source of truth for authorization logic.
---

# Write Rules

Oso provides policy building blocks for common authorization models like [role-based
access control](/guides/rbac), [attribute-based access control](/guides/attributes), and
[resource hierarchies](/guides/hierarchies).

This guide covers how to extend those building blocks by writing rules in Oso's

> policy language, Polar.  You will learn the basic structure and operations of

Reviewer 2: ```suggestion
policy language, Polar. You will learn the basic structure and operations of
```

> Polar rules, and how to write rules that refer to application classes,

Reviewer 2: ```suggestion
Polar rules and how to write rules that refer to application classes,
```

description: Learn about writing Oso policies - the source of truth for authorization logic.
---

# Write Custom Policies

> In Oso, policies are where authorization logic is stored. Policy files are
> separate from your application code, and are written in a declarative language
> called Polar (file extension `.polar`).

Reviewer 2: this is duplicated below under the `## What is a policy?` heading

```suggestion
```

Oso provides policy building blocks for common authorization models like [role-based
access control](/guides/rbac), [attribute-based access control](/guides/attributes), and
[resource hierarchies](/guides/hierarchies).

This guide covers how to extend those building blocks by writing custom
policies in Oso's policy language, Polar.
You will learn the basic structure and operations of Polar rules, and how to
write rules that refer to application classes, instances, and fields.

## What is a policy?

In Oso, policies are where authorization logic is stored. Policy files are
separate from your application code, and are written in a declarative language
called Polar (file extension `.polar`).

<!-- TODO: update enforcmeent link once docs are ported -->
Your application can query a policy using the [Enforcement APIs]({{< ref href="/guides/enforcement" lang="python" >}}) to get authorization decisions.

> The most basic component of a policy is a **rule**, which describes a logical condition that must be true in order for a query to succeed.

Reviewer 2: defining a rule as "a logical condition that must be true in order for a query to succeed" feels very abstract and not very related to authorization

Author: Taken from old language.
```suggestion
The most basic component of a policy is a **rule**, which describes who is allowed to access what resources in an application.
```

## Rules

The way you control whether an authorization query succeeds is to define and
load rules that **match** certain queries. This matching may involve
logical connectives, type checks, equality checks, variable bindings, field
lookups, method calls, arithmetic, comparisons, etc. See the [Polar Syntax
guide](polar-syntax) for a complete list of available operators.

Let's start with the basic syntax. A **rule definition** in Polar

> has a **name**, a **parameter list**, and an optional **body**.

Reviewer 2: note that we used 'arguments' instead of 'parameters' in other guides I reviewed today — I prefer 'parameters'

Author: follow up for consistency

be used in a rule body, as well as a variety of mathematical,
matching, and lookup operators.

## The Big Picture
## Query Evaluation

Before diving any deeper into the details of rules and matching,
let's take a moment to put them in context. Rules are loaded into
a **knowledge base**, a kind of specialized in-memory database that
supports **queries** by pattern matching and logical inference.

> ![Oso Architecture](guides/policies/arch.svg)

Reviewer 2: we should redo a similar diagram w/ `authorize()` instead of `is_allowed()`

Author: follow up

So far all of our examples have involved only `allow` rules
with three parameters. That's a natural place to start with Oso,
because the `authorize` function generates queries of the form
`allow(actor, action, resource)`. But you can write rules for
whatever you like, and use them from within your `allow` rules.
For instance, to use the [built-in roles features](guides/rbac), you need to
define a rule that handles roles might look like this:

> <!-- TODO: We should update these rules to reflect our current role policy semantics. -->

Reviewer 2: I think we can kill this TODO now

> <!-- TODO: We should update these rules to reflect our current role policy semantics. -->

Author: ```suggestion
```
